### PR TITLE
feat(module:slider): support nzTooltipVisible

### LINKED ISSUE #2373: slider: minimum value should always of zero position in range mode
## 重现

**Angular**

https://stackblitz.com/edit/angular-n2vcdt?file=src/app/app.component.ts

gif:

ng

**React**

https://codepen.io/cipchk/pen/wYZazP?editors=0010

gif:

react

### PR DESCRIPTION
**Pull request by wendzhue**
close #2373

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #2373

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: components/core/util/number.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  const numStr = num.toString();
+  const dotIndex = numStr.indexOf('.');
+  return dotIndex >= 0 ? numStr.length - dotIndex - 1 : 0;
+}
+
+export function ensureNumberInRange(num: number, min: number, max: number): number {
+  if (isNaN(num) || num < min) {
+    return min;
+  } else if (num > max) {
+    console.log('captured max');
```
Comment (vthinkxie) [inline components/core/util/number.ts]: no console

File: components/slider/nz-slider.spec.ts [hunk 2]
```diff
@@ -123,8 +123,12 @@ describe('NzSlider', () => {
       expect(onChangeSpy).toHaveBeenCalledTimes(1);
     });
 
-    it('should pass un-covered code testing', () => {
-      expect(sliderInstance.getValueToOffset()).toBe(sliderInstance.value);
+    it('should always display tooltip work', () => {
+
```
Comment (vthinkxie) [inline components/slider/nz-slider.spec.ts]: miss test

File: components/slider/nz-slider.spec.ts [hunk 3]
```diff
@@ -123,8 +123,12 @@ describe('NzSlider', () => {
...
+    });
+
+    it('should never display tooltip word', () => {
+
```
Comment (vthinkxie) [inline components/slider/nz-slider.spec.ts]: miss test

File: components/slider/doc/index.en-US.md [hunk 4]
```diff
@@ -28,5 +28,6 @@ To input a value in a range.
 | `[nzTipFormatter]` | Slider will pass its value to `tipFormatter`, and display its value in Tooltip, and hide Tooltip when return value is null. | `(value: number) => string` | - |
 | `[ngModel]` | The value of slider. When `range` is `false`, use `number`, otherwise, use `[number, number]` | `number｜number[]` | - |
 | `[nzVertical]` | If true, the slider will be vertical. | `boolean` | `false` |
+| `[nzTooltipVisible]` | When set to `always` tooltips are always displayed. When set to `never` they are never displayed | `default｜always｜never` | `default` |
```
Comment (vthinkxie) [inline components/slider/doc/index.en-US.md]: should be `'default'｜'always'｜'never'`

File: components/slider/doc/index.zh-CN.md [hunk 5]
```diff
@@ -29,5 +29,6 @@ title: Slider
 | `[nzTipFormatter]` | Slider 会把当前值传给 `nzTipFormatter`，并在 Tooltip 中显示 `nzTipFormatter` 的返回值，若为 null，则隐藏 Tooltip。 | `(value: number) => string` | - |
 | `[ngModel]` | 设置当前取值。当 `range` 为 `false` 时，使用 `number`，否则用 `[number, number]` | `number｜number[]` | - |
 | `[nzVertical]` | 值为 `true` 时，Slider 为垂直方向 | `boolean` | `false` |
+| `[nzTooltipVisible]` | 值为 `always` 时总是显示，值为 `never` 时在任何情况下都不显示 | `default｜always｜never` | `default` |
```
Comment (vthinkxie) [inline components/slider/doc/index.zh-CN.md]: 'default'｜'always'｜'never'

File: components/slider/nz-slider-handle.component.ts [hunk 6]
```diff
@@ -1,78 +1,126 @@
...
 @Component({
+  changeDetection    : ChangeDetectionStrategy.OnPush,
+  encapsulation      : ViewEncapsulation.None,
   selector           : 'nz-slider-handle',
   preserveWhitespaces: false,
   templateUrl        : './nz-slider-handle.component.html'
 })
-export class NzSliderHandleComponent implements OnChanges {
+export class NzSliderHandleComponent implements OnChanges, AfterViewInit, OnDestroy {
+  @ViewChild('tooltip') tooltip: NzToolTipComponent;
```
Comment (vthinkxie) [inline components/slider/nz-slider-handle.component.ts]: @ViewChild(NzToolTipComponent)

File: components/slider/nz-slider-handle.component.ts [hunk 7]
```diff
@@ -1,78 +1,126 @@
...
-      this._updateTooltipPosition(); // [For tooltip]
+    if (nzActive) {
+      if (nzActive.currentValue) {
+        this.toggleTooltip(true);
+      } else {
+        this.toggleTooltip(false);
+      }
+    }
+    if (nzTooltipVisible && !nzTooltipVisible.isFirstChange()) {
+      this.tooltip.show();
```
Comment (vthinkxie) [inline components/slider/nz-slider-handle.component.ts]: tooltip may be null

File: components/slider/nz-slider-handle.component.ts [hunk 8]
```diff
@@ -1,78 +1,126 @@
...
-  onMouseEnter($event: MouseEvent): void {
-    if (!this._slider.isDragging) {
-      this.nzActive = true;
+  ngAfterViewInit(): void {
+    if (this.nzTooltipVisible === 'always' && this.tooltip) {
+      Promise.resolve().then(() => this.tooltip.show());
     }
+
+    // NOTE: run outside of Angular for performance consideration.
+    this.ngZone.runOutsideAngular(() => {
```
Comment (vthinkxie) [inline components/slider/nz-slider-handle.component.ts]: actually, there is no need to add runOutsideAngular here

File: components/slider/nz-slider-handle.component.ts [hunk 9]
```diff
@@ -1,78 +1,126 @@
...
+    this.hovers_.unsubscribe();
+  }
+
+  private toggleTooltip(show: boolean): void {
+    if (this.nzTooltipVisible !== 'default' || !this.tooltip) {
+      return;
+    }
+
+    if (show) {
+      this.tooltip.show();
```
Comment (vthinkxie) [inline components/slider/nz-slider-handle.component.ts]: this.tooltip maybe null

Comment (wendzhue) [reply components/slider/nz-slider-handle.component.ts]:
```ts
private toggleTooltip(show: boolean): void {
    if (this.nzTooltipVisible !== 'default' || !this.tooltip) {
      return;
    }

    if (show) {
      this.tooltip.show();
    } else {
      this.tooltip.hide();
    }
  }
```

If tooltip is null the method returns ahead of this.

**Comment by wendzhue** [comment]
@vthinkxie Fixed. Please review again.

**Pull request closed without merging**